In [1]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import datasets
from sklearn import svm
print ("libraries loaded")


libraries loaded


In [4]:
iris = datasets.load_iris()
print("type(iris):", type(iris))
print(type(iris.data), "iris.data.shape:", iris.data.shape)
print(type(iris.target), "iris.target.shape:", iris.target.shape)



type(iris): <class 'sklearn.datasets.base.Bunch'>
<class 'numpy.ndarray'> iris.data.shape: (150, 4)
<class 'numpy.ndarray'> iris.target.shape: (150,)


http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
test_size : float, int, or None (default is None)
If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split. If int, represents the absolute number of test samples. If None, the value is automatically set to the complement of the train size. If train size is also None, test size is set to 0.25.

random_state : int or RandomState
Pseudo-random number generator state used for random sampling.

NB: input data has 150 rows.
split into 40% as test, 60% train.
60% x 150 = 90, 40% x 150 = 60.


In [6]:
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.4, random_state=0)
print ("X_train:", type(X_train), X_train.shape)
print ("X_test:", type(X_test), X_test.shape)
print ("y_train:", type(y_train), y_train.shape)
print ("y_test:", type(y_test), y_test.shape)

X_train: <class 'numpy.ndarray'> (90, 4)
X_test: <class 'numpy.ndarray'> (60, 4)
y_train: <class 'numpy.ndarray'> (90,)
y_test: <class 'numpy.ndarray'> (60,)


http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html
kernel : string, optional (default=’rbf’)
Specifies the kernel type to be used in the algorithm. It must be one of ‘linear’, ‘poly’, ‘rbf’, ‘sigmoid’, ‘precomputed’ or a callable. If none is given, ‘rbf’ will be used. If a callable is given it is used to pre-compute the kernel matrix from data matrices; that matrix should be an array of shape (n_samples, n_samples).

C : float, optional (default=1.0)
Penalty parameter C of the error term.

NB: see documentation for rest of parameters to tune with.


In [9]:
clf = svm.SVC(kernel='linear', C=1).fit(X_train, y_train)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test, y_test) 

<class 'sklearn.svm.classes.SVC'>


0.96666666666666667

In [10]:
#try again with a different kernel.
clf = svm.SVC(kernel='poly', C=1).fit(X_train, y_train)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test, y_test) 
#no real difference. ugh

<class 'sklearn.svm.classes.SVC'>


0.96666666666666667

In [20]:
#try again with a different kernel.
clf = svm.SVC(kernel='rbf', C=1).fit(X_train, y_train)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test, y_test) 
#different result

<class 'sklearn.svm.classes.SVC'>


0.94999999999999996

In [24]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(iris.data, iris.target, test_size=0.4, random_state=1)
print ("X_train2 = X_train:", np.array_equal(X_train2, X_train))
print ("y_train2 = y_train:", np.array_equal(y_train2, y_train))
clf = svm.SVC(kernel='rbf', C=1).fit(X_train2, y_train2)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test2, y_test2) 
#NB: different result.

X_train2 = X_train: False
y_train2 = y_train: False
<class 'sklearn.svm.classes.SVC'>


0.98333333333333328

In [25]:
X_train3, X_test3, y_train3, y_test3 = train_test_split(iris.data, iris.target, test_size=0.4, random_state=2)
print ("X_train2 = X_train3:", np.array_equal(X_train2, X_train3))
print ("y_train2 = y_train3:", np.array_equal(y_train2, y_train3))
print ("X_train3 = X_train:", np.array_equal(X_train3, X_train))
print ("y_train3 = y_train:", np.array_equal(y_train3, y_train))

clf = svm.SVC(kernel='rbf', C=1).fit(X_train3, y_train3)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test3, y_test3) 
#NB: different result.

X_train2 = X_train3: False
y_train2 = y_train3: False
X_train3 = X_train: False
y_train3 = y_train: False
<class 'sklearn.svm.classes.SVC'>


0.96666666666666667

In [26]:
X_train4, X_test4, y_train4, y_test4 = train_test_split(iris.data, iris.target, test_size=0.4, random_state=2)
print ("X_train2 = X_train3:", np.array_equal(X_train2, X_train3))
print ("y_train2 = y_train3:", np.array_equal(y_train2, y_train3))
print ("X_train3 = X_train:", np.array_equal(X_train3, X_train))
print ("y_train3 = y_train:", np.array_equal(y_train3, y_train))
print ("X_train4 = X_train3:", np.array_equal(X_train4, X_train3))
print ("y_train4 = y_train3:", np.array_equal(y_train4, y_train3))
#NB: if random_state=x then rame random selection of rows 

clf = svm.SVC(kernel='rbf', C=1).fit(X_train3, y_train3)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test3, y_test3) 
#NB: different result.

X_train2 = X_train3: False
y_train2 = y_train3: False
X_train3 = X_train: False
y_train3 = y_train: False
X_train4 = X_train3: True
y_train4 = y_train3: True
<class 'sklearn.svm.classes.SVC'>


0.96666666666666667

In [23]:
#try again with a different kernel.
clf = svm.SVC(kernel='sigmoid', C=1).fit(X_train, y_train)
print (type(clf))
#now test accuracy of the model.
clf.score(X_test, y_test) 
#NB: very low accuracy for this dataset.

<class 'sklearn.svm.classes.SVC'>


0.26666666666666666

In [ ]:
#todo: make function to repeat these predictions multipl times, record and plot score to demonstrate behaviour